drivers/power_domain/power_domain_gpio_monitor.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT power_domain_gpio_monitor
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(power_domain_gpio_monitor, CONFIG_POWER_DOMAIN_LOG_LEVEL);
+
+struct pd_gpio_monitor_config {
+	struct gpio_dt_spec power_good_gpio;
+};
+
+struct pd_gpio_monitor_data {
+	struct gpio_callback callback;
+	const struct device *dev;
+};
+
+static int pd_gpio_monitor_update_status(struct pd_gpio_monitor_data *data)
+{
+	const struct pd_gpio_monitor_config *config = data->dev->config;
+	int rc;
+
+	rc = gpio_pin_get_dt(&config->power_good_gpio);
+	if (rc < 0) {
+		LOG_WRN("Failed to read gpio logic level");
+		return rc;
+	}
+
```

Comment:
Because this logic is triggered by external events, it has the potential to break the expected chain of events.
The expectation is a driver in ACTIVE will transition through SUSPENDED before going to OFF.
This driver will instead go straight from ACTIVE to OFF.

Reply:
In monitoring power domain, there is no control over power rail, so going from ACTIVE to OFF is expected. When an interrupt is triggered, power domain notifies child nodes about the current state of rail. Monitoring power domain can't delay turning on/turning off power.

Reply:
I understand the physical transitions, but implementing it in this way greatly complicates writing device drivers.
If `PM_DEVICE_ACTION_RESUME` allocates resources and `PM_DEVICE_ACTION_SUSPEND` releases them, that callback is now required to manually check on every call which state `PM_DEVICE_ACTION_TURN_OFF` is running from, in order to determine whether the resources have been freed yet or not. This is the model we explicitly moved away from in the transition to the `PM_DEVICE_ACTION_*` model.

I also understand that doing
```
pm_device_children_action_run(data->dev, PM_DEVICE_ACTION_SUSPEND, NULL);
pm_device_children_action_run(data->dev, PM_DEVICE_ACTION_TURN_OFF, NULL);
```
is not a great solution either, as `SUSPEND` is probably doing some communication with the device, which will obviously fail here.

The best solution I can think of here is an additional action `PM_DEVICE_ACTION_OFF_FROM_ACTIVE` which directly encodes this transition and makes it simple for drivers to handle. Thoughts @gmarull ?

Reply:
I would choose to do two `PM_DEVICE_ACTION*` in sequence, it is easier to implement and does not require many changes to the existing code. Having a power bus that can change state at any time, it's expected that we don't care about shutting down devices correctly, so getting a bus error in that case is anticipated.

Reply:
The other concern is that doing this in an asynchronous manner breaks the reference counting on device runtime PM. Standard power domains get around this by only running the actions when their own reference counts go to 0, which implies the child references are also 0.  I don't know if there is a reasonable solution to that.

Reply:
Maybe custom visitor could solve this issue. By checking `pm->usage ` I could decide in what state the device should be. If `pm->usage > 0` than put a device in `RESUME`  state, otherwise put it in `SUSPEND`.

Reply:
@JordanYates Friendly ping. Could you give an opinion on if a custom visitor is a good idea?

Reply:
> Maybe custom visitor could solve this issue. By checking pm->usage  I could decide in what state the device should be. If pm->usage > 0 then put a device in RESUME state, otherwise put it in SUSPEND.

That doesn't handle the case of the reference count being `> 0` and the power domain going down. Manually setting the count to `0` will cause problems when running users try and call `pm_device_runtime_put`, and leaving it as it is will cause events to be run twice when the running users call `pm_device_runtime_put `.

Reply:
Given the asynchronous nature of monitoring power domain, I don't think there is a perfect solution. I believe that 
zeroing `pm->usage`  every time power goes down is the best possible approach. The only possible side effect that I can see is that you mentioned the `pm_device_runtime_put` can be called the ref counter set to 0:

```
if (pm->usage == 0U) {
	LOG_WRN("Unbalanced suspend");
	ret = -EALREADY;
	goto unlock;
}
```

In that case, a warning message will be logged, and the user will be informed through the return value. Additionally, the drivers themselves should return that error if the power is interrupted.

What do you think?

Reply:
@JordanYates Friendly ping

Reply:
I suspect that is the best that is possible. Whether that is acceptable to others I don't know

Reply:
I've uploaded changes as discussed in this thread. Can you take a look and give your opinion? If you believe there is somebody else that should give their feedback, feel free to add them.